Include non-obsolete location info in student profile report

regisb opened the pull request:
The student profile report that can be downloaded from the course staff
dashboard included the "location" field (which is obsolete) and the
mailing address, which is seldom completed. Here, we add the "country"
and "city" fields to the csv report. Note that this breaks compatibility
with earlier reports.

edx-webhook commented:
Thanks for the pull request, @regisb! I've created [OSPR-1155](https://openedx.atlassian.net/browse/OSPR-1155) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

nedbat commented:
@regisb thanks, but it looks like some Python test failures are due to your changes: `Country(code=None) is not JSON serializable`

regisb commented:
@nedbat Hmmmm... the fact that the we need JSON-dumpable data makes this PR more complex than I thought. It's probably the reason why the country code was not sent in the profile report in the first place :-) 
Is this feature acceptable, despite the added complexity?

nedbat commented:
@regisb I'll leave you to work that out with the T&L team.

jbarciauskas commented:
What does "break compatibility" mean in this context @regisb ?

regisb commented:
@jbarciauskas If people have written scripts that parse the user reports, e.g: Excel macros, then these scripts will probably fail because of the added columns.

pdpinch commented:
Ironically, I just had a user ask about this today. 

Can add some examples of before and after reports?

FYI @joemartis

regisb commented:
@pdpinch In the attached screenshot, I generated the report for a course with one student coming from South Georgia and the South Sandwich Islands (code=GS).

![user profiles](https://cloud.githubusercontent.com/assets/44319/13279991/a357ee1e-dadb-11e5-9c12-f5d3a34aa693.png)

And here is a screenshot of what is displayed when we click the "List enrolled students' profile information" button in the course dashboard.

![user profiles - dashboard](https://cloud.githubusercontent.com/assets/44319/13280040/f52deb26-dadb-11e5-9629-69ba28eb950b.png)

JoeMartis commented:
+1 for this PR - it would really help with our OFAC efforts.

jbarciauskas commented:
FYI @scottrish is going to shop this around for feedback on the backwards compatibility concerns

jbarciauskas commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1246,7 +1246,7 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
             'id', 'username', 'name', 'email', 'language', 'location',
             'year_of_birth', 'gender', 'level_of_education', 'mailing_address',
-            'goals'
+            'city', 'country', 'goals'
```
@regisb would it be possible to reorder this so that we don't move the "goals" position, but are just strictly adding columns to the end?

regisb replied:
Sure. But note that this query_features array does not determine the order of the columns in the Excel document. The values of PROFILE_FEATURES in instructor_analytics/basic.py do.

regisb commented:
@jbarciauskas do you wish to move the city/country fields at the end of the excel document?

jbarciauskas commented:
@regisb Yeah I think that would address the most likely compatibility issues, right?

regisb commented:
@jbarciauskas Right. I'll make the change.

scottrish commented:
@regisb Where does the city field get populated from?

regisb commented:
From the UserProfile. It is filled on a voluntary basis at sign-up or from the student preferences page. 
-- 
Régis 

Sent from my portable toaster. Please excuse the brevity. 

On 20 April 2016 13:02:15 WEST, Scott Dunn <notifications@github.com> wrote:
>@regisb Where does the city field get populated from?
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/edx/edx-platform/pull/11620#issuecomment-212397390

cahrens commented:
@regisb A city something that you have added to the sign-up and preferences page in your installation? Only Country is present on the edX site.

cahrens commented:
This PR will need tests that the new fields properly appear in the report.

regisb commented:
@cahrens Yes, we have added the 'city' field to the sign-up page (note for others: it can also be done by tweaking the REGISTRATION_EXTRA_FIELDS setting).
I also updated this PR to include the verification of the 'city' field in a unit test.

cahrens commented on lms/djangoapps/instructor_analytics/tests/test_basic.py:
```diff
@@ -106,17 +106,21 @@ def test_enrolled_students_features_username(self):
     def test_enrolled_students_features_keys(self):
-        query_features = ('username', 'name', 'email')
+        query_features = ('username', 'name', 'email', 'city')
+        for user in self.users:
```
Add country as well to the test.

regisb commented:
@cahrens added the 'country' field to the test.

cahrens commented on lms/djangoapps/instructor_analytics/basic.py:
```diff
@@ -222,6 +224,15 @@ def enrolled_students_features(course_key, features):
     if include_team_column:
         students = students.prefetch_related('teams')
 
+    def extract_attr(student, feature):
```
@regisb Do you know why this was not previously needed for "location" and "mailing_address", which aren't actually filled out (why didn't they throw an error also)?

regisb replied:
This is because 'country' is not a string: it's an instance of CountryField. When we try to serialize `student_dict` as JSON, serialization will fail for the values that do not have an acceptable type (int, str, list, etc.). That piece of code with DjangoJSONEncoder was inspired by the JSON-conversion code from django models.

cahrens commented:
This looks reasonable to me-- I'm building a sandbox to try it out.

In the meantime, @catong, I assume that documentation would need to be updated for this?

catong commented:
Yes, we'll need to update documentation. Thanks for the heads-up @cahrens. I created https://openedx.atlassian.net/browse/DOC-2946 for the doc work.

cahrens commented:
Sandbox: https://report.sandbox.edx.org/

cahrens commented:
Note that when Country is not specified, the entry is empty instead of "None". The same is true with Location, but the inconsistency that some fields (like City) say "None" is somewhat annoying.

![image](https://cloud.githubusercontent.com/assets/484484/15121655/8bcdcfd6-15e9-11e6-9083-72f254d4e0cb.png)

cahrens commented on lms/djangoapps/instructor_analytics/tests/test_basic.py:
```diff
@@ -106,17 +106,23 @@ def test_enrolled_students_features_username(self):
     def test_enrolled_students_features_keys(self):
-        query_features = ('username', 'name', 'email')
+        query_features = ('username', 'name', 'email', 'city', 'country',)
+        for user in self.users:
```
One more request-- please include a user without city and country so that we can verify the code works when those are not specified (assert the expected values).

regisb replied:
Done below.

cahrens commented:
I am asking for one more test case to cover when city and country are not present. Otherwise, I am 👍 on this PR.

@catong, I think we should wait to merge this until https://openedx.atlassian.net/browse/DOC-2946 is ready, since this change will need to be flagged in the release for compatibility reasons. We can coordinate when this one is ready to merge.

regisb commented:
@cahrens I agree that the inconsistency between "None" and "" is a bit annoying. `None` is a python-specific value, so IMHO it does make much sense to include it in a csv file. If we need to keep this value for retrocompatibility reasons, it should be relatively easy to print undefined countries as "None" by patching the `extract_attr` function. I can fix this behaviour either way, just let me know if I should make changes or create a different PR.

I'll also be adding a test for users with no city/country.

cahrens commented:
@regisb My team discussed "None" vs. "", and decided we are fine with the current implementation (inconsistency between City and Country). So no further work is needed on that end.

cahrens commented on lms/djangoapps/instructor_analytics/tests/test_basic.py:
```diff
@@ -106,17 +106,49 @@ def test_enrolled_students_features_username(self):
+
+    def test_enrolled_student_with_no_country_city(self):
+        query_features = ('username', 'city', 'country',)
+        for user in self.users:
```
Nit: I think you could just remove this block and rely on the default behavior of the fields.

cahrens commented on lms/djangoapps/instructor_analytics/tests/test_basic.py:
```diff
@@ -106,17 +106,49 @@ def test_enrolled_students_features_username(self):
+            user.profile.city = None
+            user.profile.country = None
+            user.profile.save()
+        for feature in query_features:
```
Nit: the asserts in lines 138-140 are covered in the test case above. In this test case, you could boil this down to:

```
userreports = enrolled_students_features(self.course_key, query_features)

cahrens commented:
@regisb Thanks for the additional test point. It could be simplified somewhat, but it's up to you whether or not you wish to do that. Just let me know if you'd like to make any more changes before I merge.

regisb commented:
@cahrens: Your suggestions do make sense. I took them into account and updated the PR.

cahrens commented:
jenkins run js

regisb commented:
Thanks everyone, great job!

cahrens commented:
Yes, thanks for the submission, @regisb!

cahrens commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1264,6 +1264,8 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+        'country': _('Country'),
     }
 
     if is_course_cohorted(course.id):
```
Bummer, I didn't expand below this to notice that if the course is cohorted, the cohorts column is added to the end (and the same with teams). Therefore, City and Country will not be the last two columns in the spreadsheet if either cohorts or teams are present.

We may need to revert this PR for the release. I will follow up with product.

catong replied:
Thanks @cahrens. Please tag me on any changes so I can reflect updates in the doc.

regisb replied:
What's the issue exactly?

cahrens commented:
I described it above-- City and Country will not be the last two columns if either cohorts or teams are enabled. But those are optional columns, so it's a bit funky. I need to talk with folks here on Monday about what is desired.

regisb commented:
Ok if necessary I'll have time to work on a new pr on Tuesday.
